### PR TITLE
MSVC + CUDA Support

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

**Comment by crtrott** [comment]
Just a draft

**Comment by crtrott** [comment]
Passed Windows Testing with CUDA: 

```
Test project C:/Users/ceear/source/repos/kokkos/out/build/x64-CUDA
      Start  1: KokkosCore_UnitTest_Serial1
 1/47 Test  #1: KokkosCore_UnitTest_Serial1 .................   Passed   13.01 sec
      Start  2: KokkosCore_UnitTest_Serial2
 2/47 Test  #2: KokkosCore_UnitTest_Serial2 .................   Passed    4.47 sec
      Start  3: KokkosCore_UnitTest_SerialGraph
 3/47 Test  #3: KokkosCore_UnitTest_SerialGraph .............   Passed    0.13 sec
      Start  4: KokkosCore_UnitTest_Cuda1
 4/47 Test  #4: KokkosCore_UnitTest_Cuda1 ...................   Passed   33.57 sec
      Start  5: KokkosCore_UnitTest_Cuda2
 5/47 Test  #5: KokkosCore_UnitTest_Cuda2 ...................   Passed   20.76 sec
      Start  6: KokkosCore_UnitTest_Cuda3
 6/47 Test  #6: KokkosCore_UnitTest_Cuda3 ...................   Passed   49.83 sec
      Start  7: KokkosCore_UnitTest_CudaTimingBased
 7/47 Test  #7: KokkosCore_UnitTest_CudaTimingBased .........   Passed    0.81 sec
      Start  8: KokkosCore_UnitTest_CudaInterOpInit
 8/47 Test  #8: KokkosCore_UnitTest_CudaInterOpInit .........   Passed    0.15 sec
      Start  9: KokkosCore_UnitTest_CudaInterOpStreams
 9/47 Test  #9: KokkosCore_UnitTest_CudaInterOpStreams ......   Passed    0.15 sec
      Start 10: KokkosCore_UnitTest_CudaGraph
10/47 Test #10: KokkosCore_UnitTest_CudaGraph ...............   Passed    0.16 sec
      Start 11: KokkosCore_UnitTest_Default
11/47 Test #11: KokkosCore_UnitTest_Default .................   Passed    0.29 sec
      Start 12: KokkosCore_UnitTest_PushFinalizeHook
12/47 Test #12: KokkosCore_UnitTest_PushFinalizeHook ........   Passed    0.13 sec
      Start 13: KokkosCore_UnitTest_Develop
13/47 Test #13: KokkosCore_UnitTest_Develop .................   Passed    0.12 sec
      Start 14: KokkosCore_UnitTest_LogicalSpaces
14/47 Test #14: KokkosCore_UnitTest_LogicalSpaces ...........   Passed    0.27 sec
      Start 15: KokkosCore_UnitTest_KokkosP
15/47 Test #15: KokkosCore_UnitTest_KokkosP .................   Passed    0.12 sec
      Start 16: KokkosCore_UnitTest_StackTraceTest
16/47 Test #16: KokkosCore_UnitTest_StackTraceTest ..........   Passed    0.42 sec
      Start 17: KokkosCore_UnitTest_DefaultInit_1
17/47 Test #17: KokkosCore_UnitTest_DefaultInit_1 ...........   Passed    0.16 sec
      Start 18: KokkosCore_UnitTest_DefaultInit_2
18/47 Test #18: KokkosCore_UnitTest_DefaultInit_2 ...........   Passed    0.15 sec
      Start 19: KokkosCore_UnitTest_DefaultInit_3
19/47 Test #19: KokkosCore_UnitTest_DefaultInit_3 ...........   Passed    0.13 sec
      Start 20: KokkosCore_UnitTest_DefaultInit_4
20/47 Test #20: KokkosCore_UnitTest_DefaultInit_4 ...........   Passed    0.12 sec
      Start 21: KokkosCore_UnitTest_DefaultInit_5
21/47 Test #21: KokkosCore_UnitTest_DefaultInit_5 ...........   Passed    0.16 sec
      Start 22: KokkosCore_UnitTest_DefaultInit_6
22/47 Test #22: KokkosCore_UnitTest_DefaultInit_6 ...........   Passed    0.13 sec
      Start 23: KokkosCore_UnitTest_DefaultInit_7
23/47 Test #23: KokkosCore_UnitTest_DefaultInit_7 ...........   Passed    0.13 sec
      Start 24: KokkosCore_UnitTest_DefaultInit_8
24/47 Test #24: KokkosCore_UnitTest_DefaultInit_8 ...........   Passed    0.13 sec
      Start 25: KokkosCore_UnitTest_DefaultInit_9
25/47 Test #25: KokkosCore_UnitTest_DefaultInit_9 ...........   Passed    0.14 sec
      Start 26: KokkosCore_UnitTest_DefaultInit_10
26/47 Test #26: KokkosCore_UnitTest_DefaultInit_10 ..........   Passed    0.15 sec
      Start 27: KokkosCore_UnitTest_DefaultInit_11
27/47 Test #27: KokkosCore_UnitTest_DefaultInit_11 ..........   Passed    0.15 sec
      Start 28: KokkosCore_UnitTest_DefaultInit_12
28/47 Test #28: KokkosCore_UnitTest_DefaultInit_12 ..........   Passed    0.15 sec
      Start 29: KokkosCore_UnitTest_DefaultInit_13
29/47 Test #29: KokkosCore_UnitTest_DefaultInit_13 ..........   Passed    0.15 sec
      Start 30: KokkosCore_UnitTest_DefaultInit_14
30/47 Test #30: KokkosCore_UnitTest_DefaultInit_14 ..........   Passed    0.16 sec
      Start 31: KokkosCore_UnitTest_DefaultInit_15
31/47 Test #31: KokkosCore_UnitTest_DefaultInit_15 ..........   Passed    0.15 sec
      Start 32: KokkosCore_UnitTest_DefaultInit_16
32/47 Test #32: KokkosCore_UnitTest_DefaultInit_16 ..........   Passed    0.16 sec
      Start 33: KokkosCore_UnitTest_DefaultInit_17
33/47 Test #33: KokkosCore_UnitTest_DefaultInit_17 ..........   Passed    0.15 sec
      Start 34: KokkosCore_UnitTest_DefaultInit_18
34/47 Test #34: KokkosCore_UnitTest_DefaultInit_18 ..........   Passed    0.16 sec
      Start 35: KokkosCore_IncrementalTest_CUDA
35/47 Test #35: KokkosCore_IncrementalTest_CUDA .............   Passed   35.40 sec
      Start 36: KokkosCore_IncrementalTest_SERIAL
36/47 Test #36: KokkosCore_IncrementalTest_SERIAL ...........   Passed    0.38 sec
      Start 37: KokkosCore_UnitTest_CTestDevice
37/47 Test #37: KokkosCore_UnitTest_CTestDevice .............   Passed    0.03 sec
      Start 38: KokkosCore_UnitTest_CMakePassCmdLineArgs0
38/47 Test #38: KokkosCore_UnitTest_CMakePassCmdLineArgs0 ...   Passed    0.01 sec
      Start 39: KokkosCore_PerfTestExec
39/47 Test #39: KokkosCore_PerfTestExec .....................   Passed   11.00 sec
      Start 40: KokkosCore_PerformanceTest_Atomic
40/47 Test #40: KokkosCore_PerformanceTest_Atomic ...........   Passed   17.88 sec
      Start 41: KokkosCore_PerformanceTest_Atomic_MinMax
41/47 Test #41: KokkosCore_PerformanceTest_Atomic_MinMax ....   Passed    0.15 sec
      Start 42: KokkosCore_PerformanceTest_Mempool
42/47 Test #42: KokkosCore_PerformanceTest_Mempool ..........   Passed    0.16 sec
      Start 43: KokkosCore_PerformanceTest_TaskDag
43/47 Test #43: KokkosCore_PerformanceTest_TaskDag ..........   Passed    0.01 sec
      Start 44: KokkosContainers_UnitTest_Serial
44/47 Test #44: KokkosContainers_UnitTest_Serial ............   Passed    7.21 sec
      Start 45: KokkosContainers_UnitTest_Cuda
45/47 Test #45: KokkosContainers_UnitTest_Cuda ..............   Passed   13.52 sec
      Start 46: KokkosContainers_PerformanceTest_Cuda
46/47 Test #46: KokkosContainers_PerformanceTest_Cuda .......   Passed    0.94 sec
      Start 47: KokkosAlgorithms_UnitTest
47/47 Test #47: KokkosAlgorithms_UnitTest ...................   Passed   93.68 sec
100% tests passed, 0 tests failed out of 47
Total Test time (real) = 307.51 sec
```

File: cmake/kokkos_enable_devices.cmake [hunk 1]
```diff
@@ -119,9 +119,9 @@ KOKKOS_DEVICE_OPTION(CUDA ${CUDA_DEFAULT} DEVICE "Whether to build CUDA backend"
 
 IF (KOKKOS_ENABLE_CUDA)
   GLOBAL_SET(KOKKOS_DONT_ALLOW_EXTENSIONS "CUDA enabled")
-  IF(WIN32 AND NOT KOKKOS_CXX_COMPILER_ID STREQUAL Clang)
-    GLOBAL_APPEND(KOKKOS_COMPILE_OPTIONS -x cu)
-  ENDIF()
+  #IF(WIN32 AND NOT KOKKOS_CXX_COMPILER_ID STREQUAL Clang)
+  #  GLOBAL_APPEND(KOKKOS_COMPILE_OPTIONS -x cu)
+  #ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_enable_devices.cmake:124]: Did you mean to remove?

**Comment by crtrott** [comment]
@ibaned I think this is now ready for the windows side. Passing all unit tests in Kokkos except for a couple which run into compiler bugs (explicit tests of EBO behavior and one exceeding some size limit)

File: core/src/desul/atomics/Macros.hpp [hunk 2]
```diff
@@ -20,8 +20,11 @@ SPDX-License-Identifier: (BSD-3-Clause)
 #define DESUL_HAVE_GCC_ATOMICS
 #endif
 
-#ifdef _MSC_VER
-#define DESUL_HAVE_MSVC_ATOMICS
+#if defined(_MSC_VER)  && \
+    (!defined(__CUDA_ARCH__) || !defined(__NVCC__)) && \
+    (!defined(__HIP_DEVICE_COMPILE) || !defined(__HIP_PLATFORM_HCC__)) && \
+    !defined(SYCL_LANGUAGE_VERSION)
```
Comment (dalg24) [inline core/src/desul/atomics/Macros.hpp:26]: Please elaborate on this change

File: core/src/desul/atomics/Compare_Exchange_MSVC.hpp [hunk 3]
```diff
@@ -18,6 +18,16 @@ SPDX-License-Identifier: (BSD-3-Clause)
...
 
+// Forward declare these functions. They use compare_exchange themselves
+// so the actual header file with them comes after this file is included.
+namespace Impl {
+template <typename MemoryScope>
+inline bool lock_address(void* ptr, MemoryScope ms);
+
+template <typename MemoryScope>
+void unlock_address(void* ptr, MemoryScope ms);
+}
```
Comment (dalg24) [inline core/src/desul/atomics/Compare_Exchange_MSVC.hpp:29]: I am having trouble follow that code spaghetti.
Remind me why we can't just include `Lock_Array.hpp`

Comment (crtrott) [reply core/src/desul/atomics/Compare_Exchange_MSVC.hpp:29]: as the comment above said: Lock_Array.hpp relies on compare_exchange itself, so you can't include it before defining compare_exchange.

File: core/src/traits/Kokkos_Traits_fwd.hpp [hunk 4]
```diff
@@ -45,6 +45,15 @@
 #ifndef KOKKOS_KOKKOS_TRAITS_FWD_HPP
 #define KOKKOS_KOKKOS_TRAITS_FWD_HPP
 
+// Without this the CUDA side does proper EBO while MSVC doesn't
+// leading to mismatched sizes of the driver objects (CudaParallel)
+// leading to illegal memory accesses etc on device
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/traits/Kokkos_Traits_fwd.hpp:50]: to Max, but they don't seem super interested in trying to fix this

File: containers/unit_tests/CMakeLists.txt [hunk 5]
```diff
@@ -42,6 +42,11 @@ foreach(Tag Threads;Serial;OpenMP;HPX;Cuda;HIP;SYCL)
       configure_file(${dir}/dummy.cpp ${file})
       list(APPEND UnitTestSources ${file})
     endforeach()
+    #fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj
```
[the remark this answers is not in the document]
Comment (crtrott) [reply containers/unit_tests/CMakeLists.txt:45]: no its a compiler bug

**Comment by ibaned** [comment]
Thanks @crtrott ! Let me know if you want me to test and when.

File: core/src/Kokkos_MathematicalFunctions.hpp [hunk 6]
```diff
@@ -203,6 +203,21 @@ namespace Experimental {
     return FUNC(static_cast<double>(x));                                      \
   }
 
+// isinf isnan and isinfinite don't work on windows with cuda with std::
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/Kokkos_MathematicalFunctions.hpp:206]: comment explains issue

File: cmake/kokkos_enable_devices.cmake (hunk 1 above)
Comment (crtrott) [reply cmake/kokkos_enable_devices.cmake]: done

File: cmake/kokkos_tribits.cmake [hunk 7]
```diff
@@ -244,6 +248,9 @@ MACRO(KOKKOS_ADD_TEST_EXECUTABLE ROOT_NAME)
     ""
     "SOURCES"
     ${ARGN})
+  FOREACH(source ${PARSE_SOURCES})
+    set_source_files_properties(${source} PROPERTIES LANGUAGE ${KOKKOS_COMPILE_LANGUAGE})
+  ENDFOREACH()
```
[the remark this answers is not in the document]
Comment (crtrott) [reply cmake/kokkos_tribits.cmake]: done

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 8]
```diff
@@ -142,8 +142,8 @@ class alignas(2) half_t {
   half_t() : val(0.0F) {}
 
   // Copy constructors
-  KOKKOS_DEFAULTED_FUNCTION
-  half_t(const half_t&) noexcept = default;
+  KOKKOS_FUNCTION
+  half_t(const half_t& rhs) : val(rhs.val){};
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:146]: Did that have anything to do with the `noexcept` specification?
Consider guarding the code and please add a comment.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Please open a PR on the desul side

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 8 above)
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:146]: ah shit it was guarded in my code first but the rebase was weird, gonna redo that.

File: core/src/desul/atomics/Macros.hpp (hunk 2 above)
Comment (crtrott) [reply core/src/desul/atomics/Macros.hpp:30]: done

File: core/unit_test/TestViewAPI.hpp [hunk 9]
```diff
@@ -1478,6 +1478,12 @@ class TestViewAPI {
     if (std::is_same<typename dView1::memory_space,
                      Kokkos::Experimental::OpenMPTargetSpace>::value)
       return;
+#endif
+// This test doesn't behave as expected on Windows with CUDA
```
[the remark this answers is not in the document]
Comment (crtrott) [reply core/unit_test/TestViewAPI.hpp:1483]: yeah added that.

File: core/src/Cuda/Kokkos_Cuda_Half.hpp (hunk 8 above)
Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp]: reintroduced the guard stuff

File: core/unit_test/TestViewAPI.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply core/unit_test/TestViewAPI.hpp:1483]: yeah added note

**Review by dalg24 — COMMENTED** [review]
I approve pending PR on the desul side

**Comment by crtrott** [comment]
https://github.com/desul/desul/pull/49

File: core/unit_test/TestViewAPI.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply core/unit_test/TestViewAPI.hpp:1483]: FIXME_<MSVC WITH CUDA TAG> would be better

**Review by dalg24 — APPROVED** [review]

File: core/src/Kokkos_MathematicalFunctions.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/src/Kokkos_MathematicalFunctions.hpp:206]: Should we just define `KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE` to be empty for `MSVC+CUDA` instead?

File: core/unit_test/CMakeLists.txt [hunk 10]
```diff
@@ -938,6 +938,6 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
   ARGS "one 2 THREE"
 )
 
-if (KOKKOS_ENABLE_HEADER_SELF_CONTAINMENT_TESTS AND NOT KOKKOS_HAS_TRILINOS)
+if (KOKKOS_ENABLE_HEADER_SELF_CONTAINMENT_TESTS AND NOT KOKKOS_HAS_TRILINOS AND NOT WIN32)
```
Comment (masterleinad) [inline core/unit_test/CMakeLists.txt:941]: What's the problem here?

File: core/src/Kokkos_MathematicalFunctions.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (crtrott) [reply core/src/Kokkos_MathematicalFunctions.hpp:206]: no. Generally we should call the std ones, its just that those three were overlooked by the NVCC team.

File: core/unit_test/CMakeLists.txt (hunk 10 above)
Comment (crtrott) [reply core/unit_test/CMakeLists.txt:941]: some includes are missing in windows for whatever reason and it fails inside of xstring with incomplete type issues. Somewhere we include enough of the STL to make that problem go away when you just include Kokkos_Core.hpp, but I don't think we should litter all our headerfiles with more STL includes to make the windows build happy ...

**Review by masterleinad — APPROVED** [review]
Fine with me.

**Pull request merged**